pkg/controller/ingressstate/ingress_state_controller.go

```diff
@@ -0,0 +1,247 @@
+}
+
+func NewIngressStateController(kubeInformersForTargetNamespace informers.SharedInformerFactory,
+	endpointsGetter corev1client.EndpointsGetter,
+	podsGetter corev1client.PodsGetter,
+	operatorClient v1helpers.OperatorClient,
+	targetNamespace string,
+	recorder events.Recorder,
+) *IngressStateController {
+	c := &IngressStateController{
+		endpointsGetter: endpointsGetter,
+		podGetter:       podsGetter,
+		targetNamespace: targetNamespace,
+		operatorClient:  operatorClient,
+		eventRecorder:   recorder.WithComponentSuffix("ingress-state-controller"),
+		queue:           workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "IngressStateController"),
+	}
+
+	c.cachesToSync = append(c.cachesToSync, kubeInformersForTargetNamespace.Core().V1().Pods().Informer().HasSynced)
+	c.cachesToSync = append(c.cachesToSync, kubeInformersForTargetNamespace.Core().V1().Endpoints().Informer().HasSynced)
+
+	kubeInformersForTargetNamespace.Core().V1().Pods().Informer().AddEventHandler(c.eventHandler())
+	kubeInformersForTargetNamespace.Core().V1().Endpoints().Informer().AddEventHandler(c.eventHandler())
+
+	return c
+}
+
+func (c *IngressStateController) eventHandler() cache.ResourceEventHandler {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.queue.Add(ingressStateControllerWorkQueueKey) },
+		UpdateFunc: func(old, new interface{}) { c.queue.Add(ingressStateControllerWorkQueueKey) },
+		DeleteFunc: func(obj interface{}) { c.queue.Add(ingressStateControllerWorkQueueKey) },
+	}
+}
+
+// checkPodStatus will check the target pod container status and return list of possible problems.
+func (c *IngressStateController) checkPodStatus(reference *corev1.ObjectReference) []string {
+	pod, err := c.podGetter.Pods(reference.Namespace).Get(reference.Name, metav1.GetOptions{})
+	if err != nil {
+		return []string{fmt.Sprintf("error getting pod %q: %v", reference.Name, err)}
+	}
+	result := []string{}
+	for _, containerStatus := range pod.Status.ContainerStatuses {
+		if containerStatus.Ready && containerStatus.State.Running != nil {
+			continue
+		}
+		if terminated := containerStatus.State.Terminated; terminated != nil {
+			result = append(result, fmt.Sprintf("pod %q container %q terminated with %q", pod.Name, containerStatus.Name, terminated.Message))
+		}
+		if containerStatus.RestartCount > 1 {
+			result = append(result, fmt.Sprintf("pod %q container %q restarted %d times", pod.Name, containerStatus.Name, containerStatus.RestartCount))
+		}
+		// TODO: Add more here, like we can detect "pending" pods
+	}
+	return result
+}
+
+// checkEndpointHealthz will check the health of given https://endpointIP:6443/healthz
+// TODO: this is insecure for now, but enough to tell us if we can get through
+func (c *IngressStateController) checkEndpointHealthz(endpointIP string) error {
+	// TODO: consider making this secure, however it will mean wiring the trust which must match the same trust as ingress controller
```

Comment:
@stlaz might have opinions on this

Reply:
Like I said before, the endpoint should be served using the service-ca-signed server cert, you could probably even use the CA from `/var/run/secrets/kubernetes.io/serviceaccount/ca.crt`

Reply:
afaik the service ca needs to be used since the endpoints use a serving cert.

Reply:
service-ca should be a part of the above bundle AFAIK, either should work, but I am also inclined to using the service-ca

Reply:
What mechanism is ensuring that the service ca bundle is included in the bundle provided with service accounts? It doesn't appear that the kao is requesting injection of the service ca bundle, so I'm not sure how the composition would work.

Also, if the service account bundle could be used to verify serving certs, why would the service ca operator need to support configmap bundle injection? Any pod that wanted to verify serving cert connections could just use the ca bundle supplied with its service account. If this is not supported today, maybe it should be?

Reply:
@stlaz You were on the right track - the service ca bundle is provided verbatim with every service account token as `/var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt`. I'm not sure why this shouldn't be used instead of configmap injection.